lib/ansible/modules/cloud/google/gce_bucket.py

```diff
@@ -0,0 +1,656 @@
+
+    changed = False
+    if bucket:
+        changed, versioning_status = create_bucket(module, gs, bucket)
+        if not obj:
+            if not changed:
+                module.exit_json(msg="Bucket '" + bucket + "' already exists.", changed=False, versioning=versioning_status)
+            module.exit_json(msg="Bucket created successfully", changed=changed, versioning=versioning_status)
+        else:
+            dirobj = obj if obj.endswith('/') else (obj + "/")
+
+            if key_check(module, gs, bucket, dirobj):
+                module.exit_json(msg="Bucket %s and key %s already exists." % (bucket, obj), changed=changed)
+            else:
+                create_bucket_folder(module, gs, bucket, dirobj)
+
+
+def main():
+    argument_spec = dict(
+        credentials_file=dict(type='path'),
+        service_account_email=dict(),
+        service_account_permissions=dict(type='list'),
+        project_id=dict(),
+        bucket=dict(required=True),
+        object=dict(default=None, type='path'),
+        src=dict(default=None),
+        dest=dict(default=None, type='path'),
+        expiration=dict(type='int', default=600, aliases=['expiry']),
+        mode=dict(choices=['get', 'put', 'delete', 'create', 'get_url', 'get_str'], required=True),
+        permission=dict(choices=['private', 'public-read', 'authenticated-read'], default='private'),
+        overwrite=dict(default=True, type='bool', aliases=['force']),
+        region=dict(default='US', type='str'),
+        versioning=dict(default='no', type='bool'),
+        force_delete=dict(default='no', type='bool'),
+        policy_document=dict(required=False, default=None, type='json')
+    )
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_if=[
+            ['state', 'put', ['src', 'bucket', 'object']],
+            ['state', 'get', ['dest', 'bucket', 'object']],
+        ]
+    )
+
+    if not HAS_GOOGLE_API_CORE:
+        module.fail_json(msg="Please install google-api-core to run this module.")
+
+    bucket = module.params.get('bucket')
+    obj = module.params.get('object')
+    src = module.params.get('src')
+    dest = module.params.get('dest')
+    mode = module.params.get('mode')
+    expiry = module.params.get('expiration')
+    overwrite = module.params.get('overwrite')
+    force_delete = module.params.get('force_delete')
+    service_account_permissions = module.params.get('service_account_permissions')
+
+    credentials, conn_params = get_google_cloud_credentials(module, service_account_permissions)
+    project = conn_params.get('project_id', None)
+    gs = gcp_storage_connect(module, credentials, conn_params, service_account_permissions)
+
+    if mode == 'create':
+        create_operation(module, gs, bucket, obj)
+
+    if mode == 'delete':
+        delete_operation(module, gs, bucket, obj, force_delete)
+
+    if mode == 'get':
+        if not bucket_check(module, gs, bucket) or not key_check(module, gs, bucket, obj):
```

Comment:
Please separate this into two separate error checking blocks: one for buckets and one for keys.

Reply:
Addressed this in latest commits. Please review.